src/index.tsx

```diff
@@ -43,31 +44,15 @@ export function useMutation<T extends OperationBase>(
   });
 
   const isMounted = useMounted();
+  const userConfigRef = useUpdatedRef(userConfig);
```

Comment:
I think it would be safer to have something like a `useStableValueDeep` hook that makes the ref an internal implementation detail. Something like:

```js
// TODO: Better name
function useStableValueDeep(value) {
  const valueRef = useRef(undefined);

  if (!isEqual(valueRef.current, value)) {
    valueRef.current = cloneDeep(value);
  }

  return valueRef.current;
}
```

The cloning is somewhat paranoid, but you get the idea. If we drop it, we could make this a more generic thing that takes a custom equality comparator that maybe defaults to shallow-compare instead of deep-compare.

Reply:
And then you can pass the output of `useStableValueDeep(userConfig)` into `useMemo`, and it will do what we want.

Reply:
that seems like more work for the same result? In this case we really don't ever _want_ `mutate` to change, it changing is only a consequence of needing to refresh the closure, which this avoids. thinking of `[state, setState]` where `setState` never changes regardless of `states` value or inputs

Reply:
This actually gets around the potential weirdness of stuff in concurrent mode. I believe the concern with concurrent mode here specifically is that you can be part of the way through another render cycle, or even through a partly aborted render cycle, and `someRef.current` can then take value that is not what's actually rendered to the DOM.

Reply:
IOW, callbacks that close over values cannot in general be fully stable in a concurrent-safe world.

But also I think this is a slightly neater API. We'd just do something like `value = useWhatever(value)` and then all the built-in stuff like `useMemo` and `useCallback` "just work".

Reply:
Yeah i mean i think it's a good hook, and nicer alternative to `useMemo`s limitations. I'm not sure pulling in lodash's isEqual and running it on each render is worth the cost tho, and by that I mean I just don't have any sense of the cost. Its a lot of extra work to avoid the useMemo approach in user-code, especially since it's likely that variables and configs are very likely going to be referentially different but the same value-wise on each render.

Reply:
> [...] values cannot in general be fully stable in a concurrent-safe world.

unrelated aside, you can reasonably have that but it limits your callbacks to only firing after render, which is how `useEventCallback` manages it

Reply:
Is `useCommittedRef` what we'd want here, then?

Reply:
depends, what you think the right behavior is when called immediately i guess? seems probably like the better starting point i guess.

Reply:
they're both dealing with edge cases, but it seems like "someone calling `mutate()` during render" is less likely than "concurrent mode happens". `useEffect` callbacks always fire in the order declared, right?

personally i'd just use the "stable identity" helper though :p `isEqual` should be cheap